### ISSUE TITLE
Abstract ReadOnlyCollection generates CA2227

### ISSUE THREAD

**Issue by Mikejo5000** [dotnet/roslyn-analyzers#3313]
Ported from MicrosoftDocs/visualstudio-docs#4322, opened originally by @PawcioCNCProfi

I have a setup like this:
```
/// <summary>Base class to types which define an option for command line.</summary>
public abstract class CMDOption {
	/// <summary>Specifies list of available option names that can be used as command line arguments to recognize an option.</summary>
	public abstract ReadOnlyCollection<string> names { get; protected set; }
	//...
}
```

I don't get why it shows the warning under `names`.

**Comment by Evangelink** [comment]
Technical answer: the diagnostic is reported because your property implements `ICollection`/`ICollection<T>` and you have a publicly available setter `protected set`. 

Functional answer and quoting the documentation:
> If replacing the collection is a goal, the preferred design pattern is to include a method to remove all the elements from the collection, and a method to repopulate the collection.

Frankly speaking when I see the exclusions that are made in the code for arrays, and immutable collections, I am not sure to understand why readonly collections haven't been excluded too. My point of view is that the rule should exclude all readonly/freezed/immutable collections with maybe an option to allow reporting on them as I am not 100% convinced by why arrays and immutable collections were excluded. (See conversation https://github.com/dotnet/roslyn-analyzers/issues/1900).

Tagging @mavasani

**Comment by mavasani** [comment]
It seems reasonable to exclude `ReadOnlyCollection` types from this rule.
